File: LocalPods/DialogView/DialogView/Core/Dialog.m
//
//  Dialog.m
//  TCLPlus
//
//  Created by OwenChen on 2020/8/18.
//  Copyright © 2020 TCL Electronics Holdings Ltd. All rights reserved.
//

#import "Dialog.h"

#import "UIViewController+Dialog.h"

typedef void (^OverTimeOnShowBlock)(DialogView *dialogView);


@interface DialogPriorityModel ()

@property (nonatomic, assign, getter=isDisplayed) BOOL displayed; // 是否已经显示过

@property (nonatomic, assign, getter=isNeedWaitInQueue) BOOL needWaitInQueue; // 是否需要在队列中计时

@property (nonatomic, assign, getter=isOverTimeInQueue) BOOL overTimeInQueue; // 是否在队列中已经超时

@property (nonatomic, strong) NSTimer *timerInQueue; // 队列计时器

@property (nonatomic, assign, getter=isNeedWaitOnShow) BOOL needWaitOnShow; // 是否需要在显示中计时

@property (nonatomic, assign, getter=isOverTimeOnShow) BOOL overTimeOnShow; // 是否在显示中已经超时

@property (nonatomic, strong) NSTimer *timerOnShow; // 显示计时器

@property (nonatomic, copy) OverTimeOnShowBlock timeOnShowBlock;

@end


@implementation DialogPriorityModel

- (instancetype)init {
    self = [super init];
    if (self) {
        self.showOnce = YES;
        self.endDialog = NO;
    }
    return self;
}

+ (DialogPriorityModel *)createDialogPriorityModelWithDialogView:(DialogView *)dialogView {
    DialogPriorityModel *model = [[DialogPriorityModel alloc] init];
    model.dialogView = dialogView;
    model.priorityType = DialogPriorityB;
    model.priorityValue = @(0);
    return model;
}

+ (DialogPriorityModel *)createDialogPriorityModelWithDialogView:(DialogView *)dialogView
                                                    priorityType:(DialogPriorityType)priorityType
                                                   priorityValue:(NSNumber *)priorityValue {
    DialogPriorityModel *model = [[DialogPriorityModel alloc] init];
    model.dialogView = dialogView;
    model.priorityType = priorityType;
    model.priorityValue = priorityValue ? priorityValue : @(0);
    return model;
}

+ (DialogPriorityModel *)createDialogPriorityModelWithDialogView:(DialogView *)dialogView
                                                  convenientType:(DialogConvenientType)convenientType {
    DialogPriorityModel *model = [[DialogPriorityModel alloc] init];
    model.dialogView = dialogView;

    NSInteger totalConvenientTypes = 20;
    model.priorityType = DialogPriorityA;
    model.priorityValue = @(totalConvenientTypes * 10 - convenientType * 10);

    return model;
}

- (void)setTimeInQueue:(CGFloat)timeInQueue {
    _timeInQueue = timeInQueue;

    if (_timeInQueue > 0) {
        self.needWaitInQueue = YES;
    }
}

- (void)setTimeOnShow:(CGFloat)timeOnShow {
    _timeOnShow = timeOnShow;

    if (_timeOnShow > 0) {
        self.needWaitOnShow = YES;
    }
}

- (void)setPriorityValue:(NSNumber *)priorityValue {
    if (priorityValue && [priorityValue integerValue] >= 0) {
        _priorityValue = priorityValue;
    } else {
        _priorityValue = @(0);
    }
}

- (void)startCountTimeInQueue {
    if (self.timerInQueue) {
        return;
    }

    __weak typeof(self) weakSelf = self;

    self.timerInQueue = [NSTimer scheduledTimerWithTimeInterval:self.timeInQueue repeats:YES block:^(NSTimer *_Nonnull timer) {
        weakSelf.overTimeInQueue = YES;
        [weakSelf stopCountTimeInQueue];
    }];
    [[NSRunLoop mainRunLoop] addTimer:self.timerInQueue forMode:NSRunLoopCommonModes];
}

- (void)stopCountTimeInQueue {
    if (self.timerInQueue) {
        [self.timerInQueue invalidate];
        self.timerInQueue = nil;
    }
}

- (void)startCountTimeOnShow {
    if (self.timerOnShow) {
        return;
    }

    __weak typeof(self) weakSelf = self;

    self.timerOnShow = [NSTimer scheduledTimerWithTimeInterval:self.timeOnShow repeats:YES block:^(NSTimer *_Nonnull timer) {
        weakSelf.overTimeOnShow = YES;

        if (weakSelf.timeOnShowBlock) {
            weakSelf.timeOnShowBlock(weakSelf.dialogView);
        }

        [weakSelf stopCountTimeOnShow];
    }];
    [[NSRunLoop mainRunLoop] addTimer:self.timerOnShow forMode:NSRunLoopCommonModes];
}

- (void)stopCountTimeOnShow {
    if (self.timerOnShow) {
        [self.timerOnShow invalidate];
        self.timerOnShow = nil;
    }
}

- (void)dealloc {
    [self stopCountTimeInQueue];
    [self stopCountTimeOnShow];
}

- (void)restartCountTimeOnShowWithTime:(CGFloat)timeOnShow {
    BOOL needStartCount = NO;
    if (self.timerOnShow) {
        needStartCount = YES;
        [self stopCountTimeOnShow];
    }

    self.timeOnShow = timeOnShow;

    if (needStartCount) {
        [self startCountTimeOnShow];
    }
}

@end


@interface Dialog ()

@property (nonatomic, strong) NSMutableArray<DialogPriorityModel *> *prioritySArray;
@property (nonatomic, strong) NSMutableArray<DialogPriorityModel *> *priorityAArray;
@property (nonatomic, strong) NSMutableArray<DialogPriorityModel *> *priorityBArray;
@property (nonatomic, strong) NSMutableArray<DialogPriorityModel *> *priorityCArray;
@property (nonatomic, strong) NSArray *allPriorityArray;
@property (nonatomic, strong) DialogPriorityModel *currentPriorityModel;

@end


@implementation Dialog

#pragma mark - 单例
+ (instancetype)sharedInstance {
    static Dialog *_instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[Dialog alloc] init];
        _instance.animationType = ShowAnimationFade;
        _instance.bounce = YES;
    });
    return _instance;
}

#pragma mark - Public Methods
/// 添加一个弹框
+ (DialogPriorityModel *)addDialogView:(DialogView *)dialogView
                        dialogPriority:(DialogPriorityType)priorityType
                         priorityValue:(NSNumber *)priorityValue {
    DialogPriorityModel *model = [[Dialog sharedInstance] addDialogView:dialogView dialogPriority:priorityType priorityValue:priorityValue];
    [[Dialog sharedInstance] checkCurrentPriorityModel];
    return model;
}

/// 添加一个弹框（priorityType 为 DialogPriorityB， priorityValue 为0）
+ (DialogPriorityModel *)addDialogView:(DialogView *)dialogView {
    DialogPriorityModel *model = [[Dialog sharedInstance] addDialogView:dialogView dialogPriority:DialogPriorityB priorityValue:@(0)];
    [[Dialog sharedInstance] checkCurrentPriorityModel];
    return model;
}

/// 添加多个弹框（priorityType 为 DialogPriorityB， priorityValue 为0），依次弹出
+ (NSArray<DialogPriorityModel *> *)addDialogViews:(NSArray<DialogView *> *)dialogViews {
    NSMutableArray<DialogPriorityModel *> *models = [NSMutableArray array];
    for (NSInteger i = 0; i < dialogViews.count; i++) {
        DialogPriorityModel *priorityModel = [DialogPriorityModel createDialogPriorityModelWithDialogView:[dialogViews objectAtIndex:i]
                                                                                             priorityType:DialogPriorityB
                                                                                            priorityValue:@(0)];
        [models addObject:priorityModel];
    }
    [[Dialog sharedInstance] addDialogPriorityModels:models];
    [[Dialog sharedInstance] checkCurrentPriorityModel];

    return [models copy];
}

/// 添加多个弹框模型
+ (void)addDialogPriorityModels:(NSArray<DialogPriorityModel *> *)priorityModels {
    [[Dialog sharedInstance] addDialogPriorityModels:priorityModels];
    [[Dialog sharedInstance] checkCurrentPriorityModel];
}

/// 移除所有弹框
+ (void)removeAllDialog {
    [[Dialog sharedInstance] removeAllDialog];
}

/// 移除所有弹框
/// @param exceptCurrentDialog 是否除去当前正在显示的（YES的话当前显示的弹框不移除）
+ (void)removeAllDialogWithExceptCurrentDialog:(BOOL)exceptCurrentDialog {
    [[Dialog sharedInstance] removeAllDialogWithExceptCurrentDialog:exceptCurrentDialog];
}

/// 移除指定弹框
+ (void)removeDialogWithTag:(NSInteger)tag {
    [[Dialog sharedInstance] removeDialogWithTag:tag];
}

/// 移除指定弹框
+ (void)removeDialogWithPriorityType:(DialogPriorityType)priorityType {
    [[Dialog sharedInstance] removeDialogWithPriorityType:priorityType];
}

/// 绑定弹框与控制器
+ (void)bindWithDialogView:(DialogView *)dialogView viewController:(UIViewController *)viewController {
    dialogView.boundViewController = viewController;
}

#pragma mark - Private Methods
/// 添加一个弹框
- (DialogPriorityModel *)addDialogView:(DialogView *)dialogView
                        dialogPriority:(DialogPriorityType)priorityType
                         priorityValue:(NSNumber *)priorityValue {
    DialogPriorityModel *priorityModel = [DialogPriorityModel createDialogPriorityModelWithDialogView:dialogView
                                                                                         priorityType:priorityType
                                                                                        priorityValue:priorityValue];

    [self addDialogPriorityModel:priorityModel];

    return priorityModel;
}

- (void)addDialogPriorityModel:(DialogPriorityModel *)priorityModel {
    NSMutableArray<DialogPriorityModel *> *priorityArray;
    switch (priorityModel.priorityType) {
        case DialogPriorityS:
            priorityArray = self.prioritySArray;
            break;
        case DialogPriorityA:
            priorityArray = self.priorityAArray;
            break;
        case DialogPriorityB:
            priorityArray = self.priorityBArray;
            break;
        case DialogPriorityC:
            priorityArray = self.priorityCArray;
            break;
        default:
            break;
    }

    [self insertPriorityArray:priorityArray withPriorityModel:priorityModel];
}

- (void)addDialogPriorityModels:(NSArray<DialogPriorityModel *> *)priorityModels {
    for (NSInteger i = 0; i < priorityModels.count; i++) {
        DialogPriorityModel *model = [priorityModels objectAtIndex:i];
        [self addDialogPriorityModel:model];
    }
}

- (void)insertPriorityArray:(NSMutableArray<DialogPriorityModel *> *)priorityArray withPriorityModel:(DialogPriorityModel *)priorityModel {
    @synchronized(self) {
        if (self.currentPriorityModel && self.currentPriorityModel.isEndDialog) {
            [priorityModel.dialogView closeWithAnimation:NO holdInQueue:NO];
            return;
        }

        if (!priorityArray || ![priorityArray isKindOfClass:[NSMutableArray class]] || !priorityModel) {
            return;
        }

        __weak typeof(self) weakSelf = self;
        __weak typeof(priorityModel) weakPriorityModel = priorityModel;
        [priorityModel.dialogView setUpMainReleaseFromQueueBlock:^{
            [weakSelf deletePriorityModel:weakPriorityModel];
            weakSelf.currentPriorityModel = nil;
            [weakSelf checkCurrentPriorityModel];
        }];

        priorityModel.dialogView.hidden = YES;
        priorityModel.dialogView.existPriority = YES;

        if (priorityArray.count == 0) {
            [priorityArray addObject:priorityModel];
        } else {
            for (NSInteger i = 0; i < priorityArray.count; i++) {
                DialogPriorityModel *existModel = [priorityArray objectAtIndex:i];
                if ([existModel.priorityValue floatValue] < [priorityModel.priorityValue floatValue]) {
                    [priorityArray insertObject:priorityModel atIndex:i];
                    break;
                } else if (i == priorityArray.count - 1) {
                    [priorityArray addObject:priorityModel];
                    break;
                }
            }
        }

        if (priorityModel.isNeedWaitInQueue) {
            [priorityModel startCountTimeInQueue];
        }
    }
}

- (void)checkCurrentPriorityModel {
    @synchronized(self) {
        for (NSInteger i = 0; i < self.allPriorityArray.count; i++) {
            NSMutableArray<DialogPriorityModel *> *priorityArray = [self.allPriorityArray objectAtIndex:i];
            if (priorityArray.count > 0) {
                DialogPriorityModel *priorityModel = [self firstPriorityModelInPriorityArray:priorityArray];

                if (priorityModel) {
                    if (priorityModel != self.currentPriorityModel) {
                        if (self.currentPriorityModel) {
                            self.currentPriorityModel.displayed = YES;

                            [self.currentPriorityModel.dialogView closeWithHoldInQueue:!self.currentPriorityModel.isShowOnce];

                            if (self.currentPriorityModel.isNeedWaitInQueue && !self.currentPriorityModel.isShowOnce) {
                                [self.currentPriorityModel startCountTimeInQueue];
                            }

                            if (self.currentPriorityModel.isNeedWaitOnShow && !self.currentPriorityModel.isShowOnce) {
                                [self.currentPriorityModel stopCountTimeOnShow];
                            }
                        }

                        if (priorityModel.isNeedWaitInQueue) {
                            [priorityModel stopCountTimeInQueue];
                        }

                        if (priorityModel.isNeedWaitOnShow) {
                            [priorityModel startCountTimeOnShow];

                            priorityModel.timeOnShowBlock = ^(DialogView *dialogView) {
                                [dialogView closeWithAnimation:YES holdInQueue:NO];
                            };
                        }

                        [self deleteNonexistentPriority];

                        [priorityModel.dialogView show];

                        self.currentPriorityModel = priorityModel;

                        if (priorityModel.isEndDialog) {
                            [self removeAllDialogWithExceptCurrentDialog:YES];
                        }
                    }
                }

                break;
            }
        }
    }
}

- (DialogPriorityModel *)firstPriorityModelInPriorityArray:(NSMutableArray *)priorityArray {
    @synchronized(self) {
        if (priorityArray.count > 0) {
            DialogPriorityModel *model = [priorityArray firstObject];

            BOOL passed = YES;
            if (model.dialogView.boundViewController) {
                passed = model.dialogView.boundViewController == [UIViewController windowCurrentViewController];
            }

            if ((model.isNeedWaitInQueue && model.isOverTimeInQueue) || !passed) {
                [priorityArray removeObject:model];
                return [self firstPriorityModelInPriorityArray:priorityArray];
            } else {
                return model;
            }
        } else {
            return nil;
        }
    }
}

- (void)deletePriorityModel:(DialogPriorityModel *)priorityModel {
    @synchronized(self) {
        for (NSMutableArray<DialogPriorityModel *> *array in self.allPriorityArray) {
            if ([array containsObject:priorityModel]) {
                [array removeObject:priorityModel];
            }
        }
    }
}

/// 删除除队列中之外的弹框
- (void)deleteNonexistentPriority {
    for (id object in [self keyView].subviews) {
        if ([object isKindOfClass:[DialogView class]]) {
            DialogView *view = (DialogView *)object;
            if (!view.isExistPriority) {
                [view close];
            }
        }
    }
}

- (void)removeAllDialog {
    [self removeAllDialogWithExceptCurrentDialog:NO];
}

- (UIView *)keyView {
    UIView *keyView = [UIApplication sharedApplication].keyWindow;
    if (!keyView) {
        keyView = [[[UIApplication sharedApplication] windows] lastObject];
    }
    return keyView;
}


/// 删除所有队列中的弹框
/// @param exceptCurrentDialog 是否除当前正显示的弹框（YES：不删除当前显示的弹框，NO：当前显示的弹框也删除）
- (void)removeAllDialogWithExceptCurrentDialog:(BOOL)exceptCurrentDialog {
    @synchronized(self) {
        [self deleteNonexistentPriority];

        for (NSMutableArray<DialogPriorityModel *> *models in self.allPriorityArray) {
            NSArray<DialogPriorityModel *> *needRemoves = [NSArray arrayWithArray:models];
            [models removeAllObjects];

            DialogPriorityModel *current;
            for (DialogPriorityModel *model in needRemoves) {
                if (!exceptCurrentDialog || self.currentPriorityModel != model) {
                    [model.dialogView closeWithAnimation:NO holdInQueue:NO];
                } else {
                    current = model;
                }
            }

            if (current) {
                self.currentPriorityModel = current;
                [models addObject:current];
            }
        }

        if (!exceptCurrentDialog) {
            self.currentPriorityModel = nil;
        }
    }
}

- (void)removeDialogWithTag:(NSInteger)tag {
    @synchronized(self) {
        for (NSMutableArray<DialogPriorityModel *> *models in self.allPriorityArray) {
            NSMutableArray<DialogPriorityModel *> *needRemove = [NSMutableArray array];
            for (DialogPriorityModel *model in models) {
                if (model.dialogView.tag == tag) {
                    [needRemove addObject:model];
                }
            }
            [models removeObjectsInArray:[needRemove copy]];

            for (DialogPriorityModel *model in needRemove) {
                [model.dialogView closeWithAnimation:NO holdInQueue:NO];
            }
        }

        if (self.currentPriorityModel.dialogView.tag == tag) {
            self.currentPriorityModel = nil;
        }
    }
}

- (void)removeDialogWithPriorityType:(DialogPriorityType)priorityType {
    @synchronized(self) {
        NSMutableArray<DialogPriorityModel *> *priorityArray;
        switch (priorityType) {
            case DialogPriorityS:
                priorityArray = self.prioritySArray;
                break;
            case DialogPriorityA:
                priorityArray = self.priorityAArray;
                break;
            case DialogPriorityB:
                priorityArray = self.priorityBArray;
                break;
            case DialogPriorityC:
                priorityArray = self.priorityCArray;
                break;
            default:
                break;
        }

        if (priorityArray.count > 0) {
            NSArray<DialogPriorityModel *> *needRemoves = [NSArray arrayWithArray:priorityArray];
            [priorityArray removeAllObjects];
            for (DialogPriorityModel *model in needRemoves) {
                [model.dialogView closeWithAnimation:NO holdInQueue:NO];
            }

            if (self.currentPriorityModel.priorityType == priorityType) {
                self.currentPriorityModel = nil;
            }
        }
    }
}

#pragma mark - Lazy Loading Methods
- (NSMutableArray<DialogPriorityModel *> *)prioritySArray {
    if (!_prioritySArray) {
        _prioritySArray = [NSMutableArray array];
    }
    return _prioritySArray;
}

- (NSMutableArray<DialogPriorityModel *> *)priorityAArray {
    if (!_priorityAArray) {
        _priorityAArray = [NSMutableArray array];
    }
    return _priorityAArray;
}

- (NSMutableArray<DialogPriorityModel *> *)priorityBArray {
    if (!_priorityBArray) {
        _priorityBArray = [NSMutableArray array];
    }
    return _priorityBArray;
}

- (NSMutableArray<DialogPriorityModel *> *)priorityCArray {
    if (!_priorityCArray) {
        _priorityCArray = [NSMutableArray array];
    }
    return _priorityCArray;
}

- (NSArray *)allPriorityArray {
    if (!_allPriorityArray) {
        _allPriorityArray = @[self.prioritySArray, self.priorityAArray, self.priorityBArray, self.priorityCArray];
    }
    return _allPriorityArray;
}

@end

File: LocalPods/DialogViewSwift/DialogViewSwift/Core/Dialog.swift
//
//  Dialog.swift
//  TCLHome
//
//  Created by huaizhang.chen on 2022/2/17.
//  Copyright © 2022 TCL Eagle Lab. All rights reserved.
//

import Foundation
import UIKit

class Dialog: NSObject {
    private var prioritySArray: [DialogPriorityModel] = []
    private var priorityAArray: [DialogPriorityModel] = []
    private var priorityBArray: [DialogPriorityModel] = []
    private var priorityCArray: [DialogPriorityModel] = []
    private var currentPriorityModel: DialogPriorityModel?

    static let sharedInstance = Dialog()

    // MARK: Public Methods

    /// 添加一个弹框
    /// - Parameters:
    ///   - dialogView: 弹框
    ///   - priorityType: 优先级
    ///   - priorityValue: 优先值
    /// - Returns: DialogPriorityModel
    @discardableResult
    public class func addDialog(dialogView: DialogView,
                                priorityType: DialogPriorityModel.DialogPriorityType?,
                                priorityValue: Int?) -> DialogPriorityModel {
        let model = Dialog.sharedInstance.addDialog(dialogView: dialogView,
                                                    priorityType: priorityType,
                                                    priorityValue: priorityValue)
        Dialog.sharedInstance.checkCurrentPriorityModel()
        return model
    }

    /// 添加一个弹框（priorityType 为 DialogPriorityB， priorityValue 为0）
    /// - Parameter dialogView: 弹框
    /// - Returns: DialogPriorityModel
    @discardableResult
    public class func addDialog(dialogView: DialogView) -> DialogPriorityModel {
        let model = Dialog.sharedInstance.addDialog(dialogView: dialogView,
                                                    priorityType: .priorityB,
                                                    priorityValue: 0)
        Dialog.sharedInstance.checkCurrentPriorityModel()
        return model
    }

    /// 添加多个弹框（priorityType 为 DialogPriorityB， priorityValue 为0），依次弹出
    /// - Parameter dialogViews: 弹框集合
    /// - Returns: [DialogPriorityModel]
    @discardableResult
    public class func addDialogs(dialogViews: [DialogView]) -> [DialogPriorityModel] {
        var models: [DialogPriorityModel] = []

        for dialogView in dialogViews {
            let model = Dialog.sharedInstance.addDialog(dialogView: dialogView,
                                                        priorityType: .priorityB,
                                                        priorityValue: 0)
            models.append(model)
        }

        Dialog.sharedInstance.addDialogs(priorityModels: models)
        Dialog.sharedInstance.checkCurrentPriorityModel()
        return models
    }

    /// 添加多个弹框模型
    /// - Parameter priorityModels: 弹框模型集合
    public class func addDialogs(priorityModels: [DialogPriorityModel]) {
        Dialog.sharedInstance.addDialogs(priorityModels: priorityModels)
        Dialog.sharedInstance.checkCurrentPriorityModel()
    }

    /// 移除所有弹框
    public class func removeAllDialog() {
        Dialog.sharedInstance.removeAllDialog()
    }

    /// 移除所有弹框
    /// - Parameter exceptCurrentDialog: 是否除去当前正在显示的（true 的话当前显示的弹框不移除）
    public class func removeAllDialog(exceptCurrentDialog: Bool) {
        Dialog.sharedInstance.removeAllDialog(exceptCurrentDialog: exceptCurrentDialog)
    }

    /// 移除指定弹框
    /// - Parameter tag: tag
    public class func removeDialog(tag: Int) {
        Dialog.sharedInstance.removeDialog(tag: tag)
    }

    /// 移除指定弹框
    /// - Parameter priorityType: 优先级
    public class func removeDialog(priorityType: DialogPriorityModel.DialogPriorityType) {
        Dialog.sharedInstance.removeDialog(priorityType: priorityType)
    }

    /// 绑定弹框与控制器（只会在这个控制器上显示）
    /// - Parameters:
    ///   - dialogView: 弹框
    ///   - viewController: 控制器
    public class func bindWithDialog(dialogView: DialogView, viewController: UIViewController) {
        dialogView.dataModel.boundViewController = viewController
    }

    // MARK: Private Methods

    func synchronized(_ lock: Any, closure: () -> Void) {
        objc_sync_enter(lock)
        closure()
        objc_sync_exit(lock)
    }

    private func addDialog(dialogView: DialogView,
                           priorityType: DialogPriorityModel.DialogPriorityType?,
                           priorityValue: Int?) -> DialogPriorityModel {
        let model = DialogPriorityModel.dialogPriorityModel(dialogView: dialogView,
                                                            priorityType: priorityType,
                                                            priorityValue: priorityValue)
        addDialog(priorityModel: model)
        return model
    }

    private func addDialog(priorityModel: DialogPriorityModel) {
        priorityModel.checkConfiguration()
        switch priorityModel.configuration.priorityType {
        case .priorityS: insertPriority(priorityArray: &prioritySArray, priorityModel: priorityModel)
        case .priorityA: insertPriority(priorityArray: &priorityAArray, priorityModel: priorityModel)
        case .priorityB: insertPriority(priorityArray: &priorityBArray, priorityModel: priorityModel)
        case .priorityC: insertPriority(priorityArray: &priorityCArray, priorityModel: priorityModel)
        }
    }

    private func addDialogs(priorityModels: [DialogPriorityModel]) {
        for model in priorityModels {
            addDialog(priorityModel: model)
        }
    }

    private func insertPriority(priorityArray: inout [DialogPriorityModel], priorityModel: DialogPriorityModel) {
        synchronized(self) {
            if let model = self.currentPriorityModel, model.configuration.endDialog {
                priorityModel.configuration.dialogView.closeDialogView(animation: false, stayInQueue: false)
                return
            }

            priorityModel.configuration.dialogView.setUpMainRemoveFromQueueBlock {
                self.delete(priorityModel: priorityModel)
                self.currentPriorityModel = nil
                self.checkCurrentPriorityModel()
            }

            priorityModel.configuration.dialogView.isHidden = true
            priorityModel.configuration.dialogView.dataModel.existPriority = true

            if priorityArray.isEmpty {
                priorityArray.append(priorityModel)
            } else {
                for (index, existModel) in priorityArray.enumerated() {
                    if existModel.configuration.priorityValue < priorityModel.configuration.priorityValue {
                        priorityArray.insert(priorityModel, at: index)
                        break
                    } else if index == priorityArray.count - 1 {
                        priorityArray.append(priorityModel)
                        break
                    }
                }
            }

            if priorityModel.control.needWaitInQueue {
                priorityModel.startCountTimeInQueue()
            }
        }
    }

    private func checkCurrentPriorityModel() {
        synchronized(self) {
            if !self.prioritySArray.isEmpty {
                self.checkCurrentPriorityModel(priorityArray: &self.prioritySArray)
            } else if !self.priorityAArray.isEmpty {
                self.checkCurrentPriorityModel(priorityArray: &self.priorityAArray)
            } else if !self.priorityBArray.isEmpty {
                self.checkCurrentPriorityModel(priorityArray: &self.priorityBArray)
            } else if !self.priorityCArray.isEmpty {
                self.checkCurrentPriorityModel(priorityArray: &self.priorityCArray)
            }
        }
    }

    private func checkCurrentPriorityModel(priorityArray: inout [DialogPriorityModel]) {
        synchronized(self) {
            if let priorityModel = firstPriorityModel(priorityArray: &priorityArray), priorityModel !== currentPriorityModel {
                if let currentModel = currentPriorityModel {
                    currentModel.control.displayed = true

                    currentModel.configuration.dialogView.closeDialogView(stayInQueue: !currentModel.configuration.showOnce)

                    if currentModel.control.needWaitInQueue, !currentModel.configuration.showOnce {
                        currentModel.startCountTimeInQueue()
                    }

                    if currentModel.control.needWaitOnShow, !currentModel.configuration.showOnce {
                        currentModel.stopCountTimeOnShow()
                    }
                }

                if priorityModel.control.needWaitInQueue {
                    priorityModel.stopCountTimeInQueue()
                }

                if priorityModel.control.needWaitOnShow {
                    priorityModel.startCountTimeOnShow()

                    priorityModel.control.timeOnShowBlock = { (_ dialogView: DialogView?) in
                        if let view = dialogView {
                            view.closeDialogView(animation: true, stayInQueue: false)
                        }
                    }
                }

                deleteNotExistPriority()
                priorityModel.configuration.dialogView.showDialogView()
                currentPriorityModel = priorityModel

                if priorityModel.configuration.endDialog {
                    removeAllDialog(exceptCurrentDialog: true)
                }
            }
        }
    }

    private func firstPriorityModel(priorityArray: inout [DialogPriorityModel]) -> DialogPriorityModel? {
        if !priorityArray.isEmpty, let model = priorityArray.first {
            var passed = true
            if let bound = model.configuration.dialogView.dataModel.boundViewController, let current = UIWindow().elk.currentViewController {
                passed = bound == current
            }

            if (model.control.needWaitInQueue && model.control.overTimeInQueue) || !passed {
                priorityArray.removeFirst()
                return firstPriorityModel(priorityArray: &priorityArray)
            } else {
                return model
            }
        } else {
            return nil
        }
    }

    private func delete(priorityModel: DialogPriorityModel) {
        synchronized(self) {
            self.prioritySArray.removeAll(where: { $0 === priorityModel })
            self.priorityAArray.removeAll(where: { $0 === priorityModel })
            self.priorityBArray.removeAll(where: { $0 === priorityModel })
            self.priorityCArray.removeAll(where: { $0 === priorityModel })
        }
    }

    private func deleteNotExistPriority() {
        if let view = DialogView.keyView(), let dialogView = view as? DialogView {
            if !dialogView.dataModel.existPriority {
                dialogView.closeDialogView()
            }
        }
    }

    private func removeAllDialog() {
        removeAllDialog(exceptCurrentDialog: false)
    }

    private func removeAllDialog(exceptCurrentDialog: Bool) {
        synchronized(self) {
            self.deleteNotExistPriority()

            self.removeAllDialog(priorityArray: &self.prioritySArray, exceptCurrentDialog: exceptCurrentDialog)
            self.removeAllDialog(priorityArray: &self.priorityAArray, exceptCurrentDialog: exceptCurrentDialog)
            self.removeAllDialog(priorityArray: &self.priorityBArray, exceptCurrentDialog: exceptCurrentDialog)
            self.removeAllDialog(priorityArray: &self.priorityCArray, exceptCurrentDialog: exceptCurrentDialog)

            if !exceptCurrentDialog {
                self.currentPriorityModel = nil
            }
        }
    }

    private func removeAllDialog(priorityArray: inout [DialogPriorityModel], exceptCurrentDialog: Bool) {
        let needRemoves: [DialogPriorityModel] = priorityArray
        priorityArray.removeAll()

        var current: DialogPriorityModel?
        for model in needRemoves {
            if !exceptCurrentDialog || currentPriorityModel !== model {
                model.configuration.dialogView.closeDialogView(animation: false, stayInQueue: false)
            } else {
                current = model
            }
        }

        if let currentModel = current {
            currentPriorityModel = current
            priorityArray.append(currentModel)
        }
    }

    private func removeDialog(tag: Int) {
        synchronized(self) {
            self.removeDialog(tag: tag, priorityArray: &self.prioritySArray)
            self.removeDialog(tag: tag, priorityArray: &self.priorityAArray)
            self.removeDialog(tag: tag, priorityArray: &self.priorityBArray)
            self.removeDialog(tag: tag, priorityArray: &self.priorityCArray)

            if let model = self.currentPriorityModel, model.configuration.dialogView.tag == tag {
                self.currentPriorityModel = nil
            }
        }
    }

    private func removeDialog(tag: Int, priorityArray: inout [DialogPriorityModel]) {
        synchronized(self) {
            var needRemoves: [DialogPriorityModel] = []
            for model in priorityArray where model.configuration.dialogView.tag == tag {
                needRemoves.append(model)
            }

            priorityArray.removeAll(where: { $0.configuration.dialogView.tag == tag })
        }
    }

    private func removeDialog(priorityType: DialogPriorityModel.DialogPriorityType) {
        synchronized(self) {
            switch priorityType {
            case .priorityS: removeDialog(priorityArray: &prioritySArray)
            case .priorityA: removeDialog(priorityArray: &priorityAArray)
            case .priorityB: removeDialog(priorityArray: &priorityBArray)
            case .priorityC: removeDialog(priorityArray: &priorityCArray)
            }

            if self.currentPriorityModel?.configuration.priorityType == priorityType {
                self.currentPriorityModel = nil
            }
        }
    }

    private func removeDialog(priorityArray: inout [DialogPriorityModel]) {
        synchronized(self) {
            let needRemoves: [DialogPriorityModel] = priorityArray
            priorityArray.removeAll()

            for model in needRemoves {
                model.configuration.dialogView.closeDialogView(animation: false, stayInQueue: false)
            }
        }
    }
}
